src/lib.rs

```diff
@@ -144,6 +144,21 @@ macro_rules! schema {
             fn incomplete_write(&self) -> bool {
                 self.incomplete_write
             }
+
+            fn compact_log(&self) -> Result<(), $crate::errors::Error> {
+                $(let ($table_name, writer) = self.$table_name.begin_transaction()?;)*
+
+                let mut data = vec![];
+                $(
+                    for (key, val) in $table_name.get_all() {
+                        data.push(helper_log::$table_name::insert(key, val));
+                    }
+                )*
+
+                writer.compact_log(data)?;
+
+                Ok(())
+            }
```

Comment:
I don't think it makes too much sense for this to live on the `Reader` trait. I figured that trait was for all the startup logic the database does.

I think you should define your own `LogCompacter` trait and attach it to that (see `Transaction` for inspiration).

This is probably also where [this logic](https://github.com/Parth/hmdb/issues/5) can live.